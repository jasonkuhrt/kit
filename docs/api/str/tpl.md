# Str.Tpl

_Str_ / **Tpl**

## Import

::: code-group

```typescript [Namespace]
import { Str } from '@wollybeard/kit'

// Access via namespace
Str.Tpl.someFunction()
```

```typescript [Barrel]
import * as Str from '@wollybeard/kit/str'

// Access via namespace
Str.Tpl.someFunction()
```

:::

## Template

### <span style="opacity: 0.6; font-weight: normal; font-size: 0.85em;">`[T]`</span> `Tpl`

```typescript
type Tpl = TemplateStringsArray
```

<SourceLink href="https://github.com/jasonkuhrt/kit/blob/main/./src/domains/str/tpl/tpl.ts#L19" />

Convenience re-export of the built-in TemplateStringsArray type. Contains the string parts of a tagged template literal along with a raw property.

**Examples:**

```typescript twoslash
// @noErrors
import { Str } from '@wollybeard/kit/str'
// ---cut---
function tag(strings: Str.Tpl.Tpl.Array, ...values: unknown[]) {
  // strings is TemplateStringsArray
  // strings[0] = "Hello "
  // strings[1] = "!"
  // strings.raw contains raw string values
}
tag`Hello ${name}!`
```

### <span style="opacity: 0.6; font-weight: normal; font-size: 0.85em;">`[F]`</span> `is`

```typescript
(value: unknown) => value is TemplateStringsArray
```

<SourceLink href="https://github.com/jasonkuhrt/kit/blob/main/./src/domains/str/tpl/tpl.ts#L38" />

Type guard to check if a value is a TemplateStringsArray. Used to detect when a function is called as a tagged template literal.

**Examples:**

```typescript twoslash
// @noErrors
import { Str } from '@wollybeard/kit/str'
// ---cut---
function tag(...args: unknown[]) {
  if (isArray(args[0])) {
    // Called as tag`template`
  } else {
    // Called as tag()
  }
}
```

### <span style="opacity: 0.6; font-weight: normal; font-size: 0.85em;">`[T]`</span> `CallInput`

```typescript
type CallInput = [Tpl, ...unknown[]]
```

<SourceLink href="https://github.com/jasonkuhrt/kit/blob/main/./src/domains/str/tpl/tpl.ts#L57" />

Tagged template literal arguments tuple. First element is the template strings array, followed by interpolated values.

**Examples:**

```typescript twoslash
// @noErrors
import { Str } from '@wollybeard/kit/str'
// ---cut---
function tag(...args: unknown[]) {
  if (isArgs(args)) {
    const [strings, ...values] = args
    // Process template literal
  }
}
tag`Hello ${name}!`
```

### <span style="opacity: 0.6; font-weight: normal; font-size: 0.85em;">`[F]`</span> `isCallInput`

```typescript
(value: unknown) => value is CallInput
```

<SourceLink href="https://github.com/jasonkuhrt/kit/blob/main/./src/domains/str/tpl/tpl.ts#L75" />

Type guard to check if function arguments are from a tagged template literal.

**Examples:**

```typescript twoslash
// @noErrors
import { Str } from '@wollybeard/kit/str'
// ---cut---
function tag(...args: unknown[]) {
  if (isArgs(args)) {
    const [strings, ...values] = args
    // Process as template literal
  }
}
tag`Hello ${name}!`
```

### <span style="opacity: 0.6; font-weight: normal; font-size: 0.85em;">`[F]`</span> `normalizeCall`

```typescript
;((callInput: CallInput) => Call)
```

<SourceLink href="https://github.com/jasonkuhrt/kit/blob/main/./src/domains/str/tpl/tpl.ts#L102" />

Parse tagged template literal arguments into structured parts and values.

**Examples:**

```typescript twoslash
// @noErrors
import { Str } from '@wollybeard/kit/str'
// ---cut---
function tag(...args: unknown[]) {
  if (isArgs(args)) {
    const { parts, values } = parse(args)
    // parts[0] = "Hello "
    // parts[1] = "!"
    // values[0] = name
  }
}
tag`Hello ${name}!`
```

### <span style="opacity: 0.6; font-weight: normal; font-size: 0.85em;">`[F]`</span> `renderWith`

```typescript
;((mapper: (value: unknown) => string) => (callInput: CallInput) => string)
```

<SourceLink href="https://github.com/jasonkuhrt/kit/blob/main/./src/domains/str/tpl/tpl.ts#L125" />

Render tagged template literal arguments using a custom value renderer.

**Examples:**

```typescript twoslash
// @noErrors
import { Str } from '@wollybeard/kit/str'
// ---cut---
// Custom renderer for JSON values
// [!code word:renderWith:1]
// [!code word:stringify:1]
const renderJson = Str.Tpl.renderWith(v => JSON.stringify(v))
function tag(...args: unknown[]) {
  if (isArgs(args)) return renderJson(args)
}
tag`Value: ${{ foo: 'bar' }}` // "Value: {\"foo\":\"bar\"}"

// Custom renderer that prefixes values
// [!code word:renderWith:1]
const renderPrefixed = Str.Tpl.renderWith(v => `[${v}]`)
```

### <span style="opacity: 0.6; font-weight: normal; font-size: 0.85em;">`[C]`</span> `render`

```typescript
;((callInput: CallInput) => string)
```

<SourceLink href="https://github.com/jasonkuhrt/kit/blob/main/./src/domains/str/tpl/tpl.ts#L150" />

Render tagged template literal arguments to a string. Interpolated values are converted using plain String() coercion.

**Examples:**

```typescript twoslash
// @noErrors
import { Str } from '@wollybeard/kit/str'
// ---cut---
function tag(...args: unknown[]) {
  if (isArgs(args)) {
    // [!code word:render:1]
    return Str.Tpl.render(args)
  }
}
tag`Hello ${name}!` // "Hello World!"
tag`Count: ${42}` // "Count: 42"
```

### <span style="opacity: 0.6; font-weight: normal; font-size: 0.85em;">`[F]`</span> `passthrough`

```typescript
;((strings: TemplateStringsArray, ...values: unknown[]) => string)
```

<SourceLink href="https://github.com/jasonkuhrt/kit/blob/main/./src/domains/str/tpl/tpl.ts#L166" />

A passthrough tagged template literal that returns the interpolated string as-is. Useful for semantic clarity in code without any processing.

**Examples:**

```typescript twoslash
// @noErrors
import { Str } from '@wollybeard/kit/str'
// ---cut---
const template = passthrough
const message = template`Hello ${name}, you have ${count} items.`
// Result: "Hello Alice, you have 5 items."
```

### <span style="opacity: 0.6; font-weight: normal; font-size: 0.85em;">`[F]`</span> `dedent`

```typescript
;((strings: TemplateStringsArray, ...values: unknown[]) => string)
```

<SourceLink href="https://github.com/jasonkuhrt/kit/blob/main/./src/domains/str/tpl/tpl.ts#L209" />

Tagged template literal that removes common indentation from all lines. Automatically indents multi-line interpolated values to match their context.

Uses the raw template strings to preserve escape sequences (e.g., \n stays as backslash-n). Trims leading and trailing blank lines from the result.

**Examples:**

```typescript twoslash
// @noErrors
import { Str } from '@wollybeard/kit/str'
// ---cut---
const code = dedent`
  function greet() {
    console.log('Hello')
  }
`
// Result: "function greet() {\n  console.log('Hello')\n}"
```

```typescript twoslash
// @noErrors
import { Str } from '@wollybeard/kit/str'
// ---cut---
// Multi-line values are auto-indented
const inner = 'line1\nline2'
const code = dedent`
  outer:
    ${inner}
`
// Result: "outer:\n  line1\n  line2"
```

```typescript twoslash
// @noErrors
import { Str } from '@wollybeard/kit/str'
// ---cut---
// Escape sequences are preserved
const path = dedent`
  C:\Users\name\Documents
`
// Result: "C:\\Users\\name\\Documents" (backslashes preserved)
```

### <span style="opacity: 0.6; font-weight: normal; font-size: 0.85em;">`[T]`</span> `HighlightTag`

```typescript
type HighlightTag = typeof passthrough
```

<SourceLink href="https://github.com/jasonkuhrt/kit/blob/main/./src/domains/str/tpl/tpl.ts#L250" />

Type for a tagged template literal function used for syntax highlighting.

### <span style="opacity: 0.6; font-weight: normal; font-size: 0.85em;">`[C]`</span> `highlight`

```typescript
{
  ts: ;
  ;((strings: TemplateStringsArray, ...values: unknown[]) => string)
  js: ;
  ;((strings: TemplateStringsArray, ...values: unknown[]) => string)
  html: ;
  ;((strings: TemplateStringsArray, ...values: unknown[]) => string)
  css: ;
  ;((strings: TemplateStringsArray, ...values: unknown[]) => string)
  sql: ;
  ;((strings: TemplateStringsArray, ...values: unknown[]) => string)
  json: ;
  ;((strings: TemplateStringsArray, ...values: unknown[]) => string)
  yaml: ;
  ;((strings: TemplateStringsArray, ...values: unknown[]) => string)
  yml: ;
  ;((strings: TemplateStringsArray, ...values: unknown[]) => string)
  graphql: ;
  ;((strings: TemplateStringsArray, ...values: unknown[]) => string)
  gql: ;
  ;((strings: TemplateStringsArray, ...values: unknown[]) => string)
  iso: ;
  ;((strings: TemplateStringsArray, ...values: unknown[]) => string)
}
```

<SourceLink href="https://github.com/jasonkuhrt/kit/blob/main/./src/domains/str/tpl/tpl.ts#L285" />

Object containing language-specific template tag functions for syntax highlighting. Each property is a tagged template function that provides editor syntax highlighting for that language (when supported by the editor).

Implemented as a Proxy that returns the same passthrough function for all properties, allowing destructuring and property access to work seamlessly.

Supported languages are based on common supported editor injection patterns:

**Examples:**

```typescript twoslash
// @noErrors
import { Str } from '@wollybeard/kit'

// [!code word:highlight:1]
const { ts, html, sql } = Str.Tpl.highlight

const code = ts`
  export const add = (a: number, b: number) => a + b
` // Gets TypeScript syntax highlighting in editor

const markup = html`
  <div class="container">Hello</div>
` // Gets HTML syntax highlighting

const query = sql`
  SELECT * FROM users WHERE id = ${userId}
` // Gets SQL syntax highlighting
```

## Other

### <span style="opacity: 0.6; font-weight: normal; font-size: 0.85em;">`[I]`</span> `Call`

```typescript
interface Call {
  template: Tpl
  args: unknown[]
}
```

<SourceLink href="https://github.com/jasonkuhrt/kit/blob/main/./src/domains/str/tpl/tpl.ts#L79" />
