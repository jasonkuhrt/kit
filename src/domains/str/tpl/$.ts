export * as Tpl from './tpl.js'
