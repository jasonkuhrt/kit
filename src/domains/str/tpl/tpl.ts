/**
 * Convenience re-export of the built-in TemplateStringsArray type.
 * Contains the string parts of a tagged template literal along with a `raw` property.
 * @category Template
 * @example
 * ```typescript
 * function tag(strings: Tpl.Array, ...values: unknown[]) {
 *   // strings is TemplateStringsArray
 *   // strings[0] = "Hello "
 *   // strings[1] = "!"
 *   // strings.raw contains raw string values
 * }
 * tag`Hello ${name}!`
 * ```
 */
export type Tpl = TemplateStringsArray

/**
 * Type guard to check if a value is a TemplateStringsArray.
 * Used to detect when a function is called as a tagged template literal.
 * @category Template
 * @param value - Value to check
 * @returns True if value is a TemplateStringsArray
 * @example
 * ```typescript
 * function tag(...args: unknown[]) {
 *   if (isArray(args[0])) {
 *     // Called as tag`template`
 *   } else {
 *     // Called as tag()
 *   }
 * }
 * ```
 */
export const is = (value: unknown): value is Tpl => {
  return Array.isArray(value) && value.length > 0 && value[0] instanceof Object && `raw` in (value[0] as any)
}

/**
 * Tagged template literal arguments tuple.
 * First element is the template strings array, followed by interpolated values.
 * @category Template
 * @example
 * ```typescript
 * function tag(...args: unknown[]) {
 *   if (isArgs(args)) {
 *     const [strings, ...values] = args
 *     // Process template literal
 *   }
 * }
 * tag`Hello ${name}!`
 * ```
 */
export type CallInput = [Tpl, ...unknown[]]

/**
 * Type guard to check if function arguments are from a tagged template literal.
 * @category Template
 * @param value - Function arguments to check
 * @returns True if args are tagged template literal arguments
 * @example
 * ```typescript
 * function tag(...args: unknown[]) {
 *   if (isArgs(args)) {
 *     const [strings, ...values] = args
 *     // Process as template literal
 *   }
 * }
 * tag`Hello ${name}!`
 * ```
 */
export const isCallInput = (value: unknown): value is CallInput => {
  return Array.isArray(value) && is(value[0])
}

export interface Call {
  template: Tpl
  args: unknown[]
}

/**
 * Parse tagged template literal arguments into structured parts and values.
 * @category Template
 * @param callInput - Tagged template literal arguments
 * @returns Object with parts (TemplateStringsArray) and values (unknown[])
 * @example
 * ```typescript
 * function tag(...args: unknown[]) {
 *   if (isArgs(args)) {
 *     const { parts, values } = parse(args)
 *     // parts[0] = "Hello "
 *     // parts[1] = "!"
 *     // values[0] = name
 *   }
 * }
 * tag`Hello ${name}!`
 * ```
 */
export const normalizeCall = (callInput: CallInput): Call => {
  const [template, ...args] = callInput
  return { template, args }
}

/**
 * Render tagged template literal arguments using a custom value renderer.
 * @category Template
 * @param mapper - Function to convert interpolated values to strings
 * @returns Function that takes template args and returns rendered string
 * @example
 * ```typescript
 * // Custom renderer for JSON values
 * const renderJson = renderWith(v => JSON.stringify(v))
 * function tag(...args: unknown[]) {
 *   if (isArgs(args)) return renderJson(args)
 * }
 * tag`Value: ${{ foo: 'bar' }}` // "Value: {\"foo\":\"bar\"}"
 *
 * // Custom renderer that prefixes values
 * const renderPrefixed = renderWith(v => `[${v}]`)
 * ```
 */
export const renderWith = (mapper: (value: unknown) => string) => (callInput: CallInput): string => {
  const call = normalizeCall(callInput)
  return call.template.reduce(
    (result, part, i) => `${result}${part}${i in call.args ? mapper(call.args[i]) : ``}`,
    ``,
  )
}

/**
 * Render tagged template literal arguments to a string.
 * Interpolated values are converted using plain `String()` coercion.
 * @category Template
 * @param args - Tagged template literal arguments
 * @returns Rendered string with all parts and values concatenated
 * @example
 * ```typescript
 * function tag(...args: unknown[]) {
 *   if (isArgs(args)) {
 *     return render(args)
 *   }
 * }
 * tag`Hello ${name}!` // "Hello World!"
 * tag`Count: ${42}` // "Count: 42"
 * ```
 */
export const render = renderWith(String)
