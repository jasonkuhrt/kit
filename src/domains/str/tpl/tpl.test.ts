import { Test } from '#test'
import * as Tpl from './tpl.js'

const getTpl = (...args: Tpl.CallInput) => args[0]
const getCallInput = (...args: Tpl.CallInput) => args

// dprint-ignore
Test.on(Tpl.is)
  .cases(
    // True cases
    [[getTpl`hello`],           true],
    [[getTpl`hello ${'x'}`],    true],

    // False cases
    [[[],                       false]],
    [[([`a`, `b`] as any),      false]],
    [[({ raw: [`test`] }),      false]],
    [[null,                     false]],
    [[42,                       false]],
  )
  .test()

// dprint-ignore
Test.on(Tpl.isCallInput)
  .cases(
    // True cases
    [[getCallInput`test`,            true]],
    [[getCallInput`hello ${1}`,      true]],

    // False cases
    [[null,                     false]],
    [[[],                       false]],
    [[[[`not`, `template`]],    false]],
  )
  .test()

// dprint-ignore
Test.describe(`passthrough`)
  .inputType<string>()
  .outputType<string>()
  .cases(
    [Tpl.passthrough`Hello ${`World`}!`,                                `Hello World!`],
    [Tpl.passthrough`Sum: ${1} + ${2} = ${3}`,                          `Sum: 1 + 2 = 3`],
    [Tpl.passthrough`\n      export const foo = ${42}\n    `,           `\n      export const foo = 42\n    `],
  )
  .test()

// dprint-ignore
Test.describe(`dedent`)
  .inputType<string>()
  .outputType<string>()
  .cases(
    // Remove common indentation, preserve relative indentation
    [Tpl.dedent`
      line 1
      line 2
        indented
    `, `line 1\nline 2\n  indented`],

    // Trim leading/trailing blank lines
    [Tpl.dedent`

      content

    `, `content`],

    // Single-line interpolation
    [Tpl.dedent`
      Hello ${`world`}
      How are you?
    `, `Hello world\nHow are you?`],

    // Multi-line value auto-indentation (key feature for dindist #9)
    [Tpl.dedent`
      outer:
        ${`line1\nline2\nline3`}
    `, `outer:\n  line1\n  line2\n  line3`],

    // Preserve internal blank lines
    [Tpl.dedent`
      line 1

      line 3
    `, `line 1\n\nline 3`],

    // Edge case: empty
    [Tpl.dedent``, ``],
  )
  .test()

// dprint-ignore
Test.describe(`highlight`)
  .inputType<string>()
  .outputType<string>()
  .cases(
    [Tpl.highlight.ts`const x = ${100}`,    `const x = 100`],
    [Tpl.highlight.html`test`,              `test`],
    [Tpl.highlight.sql`test`,               `test`],
  )
  .test()

// Regression test for dindist issue #9: nested dedent calls with function interpolation
// This validates that multi-line interpolated values get auto-indented to match context
// https://github.com/jasonkuhrt-archive/dindist/issues/9
Test.describe(`dedent nested calls`)
  .inputType<string>()
  .outputType<string>()
  .cases(
    [
      (() => {
        const intro = () =>
          Tpl.dedent`
        * intro line
      `
        const body = Tpl.dedent`
        * body line 1
        * body line 2
      `
        const outro = () =>
          Tpl.dedent`
        * outro line
      `

        return Tpl.dedent`
        /**
         ${intro()}
          *
          ${body}
          *
          ${outro()}
          */
      `
      })(),
      `/**
 * intro line
 *
 * body line 1
 * body line 2
 *
 * outro line
 */`,
    ],
  )
  .test()
