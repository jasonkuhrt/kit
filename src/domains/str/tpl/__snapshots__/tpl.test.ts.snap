// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`is([ 'a', 'b' ], false) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
[ 'a', 'b' ]
────────────────────────────────────────────────────
false
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is([], false) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
[]
────────────────────────────────────────────────────
false
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is({ raw: [ 'test' ] }, false) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  raw: [ 'test' ]
}
────────────────────────────────────────────────────
false
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is(42, false) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
42
────────────────────────────────────────────────────
false
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is(null, false) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
null
────────────────────────────────────────────────────
false
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`isCallInput([ [ 'hello ', '' ], 1 ], true) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
[ [ 'hello ', '' ], 1 ]
────────────────────────────────────────────────────
true
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`isCallInput([ [ 'not', 'template' ] ], false) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
[ [ 'not', 'template' ] ]
────────────────────────────────────────────────────
false
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`isCallInput([ [ 'test' ] ], true) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
[ [ 'test' ] ]
────────────────────────────────────────────────────
true
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`isCallInput([], false) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
[]
────────────────────────────────────────────────────
false
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`isCallInput(null, false) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
null
────────────────────────────────────────────────────
false
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;
